### PR TITLE
[feature-wip][UDF][DIP-1] Support variable-size input and output for Java UDF

### PR DESCRIPTION
**Pull request by Gabriel39**
# Proposed changes

Issue Number: [#8389](https://github.com/apache/incubator-doris/issues/8389)

## Problem Summary:

This feature is propsoed in [DIP-1](https://cwiki.apache.org/confluence/display/DORIS/DSIP-001%3A+Java+UDF). This PR support variable-length input and output Java UDF.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/function_java_udf.h [hunk 1]
```diff
@@ -76,27 +76,37 @@ class JavaFunctionCall : public IFunctionBase {
...
+
     struct JniContext {
         JavaFunctionCall* parent = nullptr;
 
         jobject executor = nullptr;
 
         int64_t input_values_buffer_ptr;
         int64_t input_nulls_buffer_ptr;
-        int64_t input_byte_offsets_ptr;
+        int64_t input_offsets_ptrs;
```
Comment (HappenLee) [inline be/src/vec/functions/function_java_udf.h:91]: maybe use unique_ptr to manage the mem is better， the int64_t only use in thrift struct

Comment (Gabriel39) [reply be/src/vec/functions/function_java_udf.h]: done

File: be/src/vec/functions/function_java_udf.h [hunk 2]
```diff
@@ -76,27 +76,36 @@ class JavaFunctionCall : public IFunctionBase {
...
 
         jobject executor = nullptr;
 
-        int64_t input_values_buffer_ptr;
-        int64_t input_nulls_buffer_ptr;
-        int64_t input_byte_offsets_ptr;
-        int64_t output_value_buffer;
-        int64_t output_null_value;
-        int64_t batch_size_ptr;
+        std::unique_ptr<int64_t> input_values_buffer_ptr;
```
Comment (HappenLee) [inline be/src/vec/functions/function_java_udf.h:89]: use `int64_t[]` not `int64_t`

**Comment by Gabriel39** [comment]
I tested performance on SSB. Since Vectorized engine doesn't support native UDF, I ran Java UDF and native UDF on vectorized engine and row-based engine separately. And a simple `AddOne` UDF for non-null int type is used for both Java UDF and native UDF.

I got results below,
`set enable_vectorized_engine=true;
select java_udf_add_one(c_custkey) from customer;
3000000 rows in set (49.02 sec)

set enable_vectorized_engine=false;
select native_udf_add_one(c_custkey) from customer;
3000000 rows in set (49.77 sec)`

which means there is no obvious performance gap between Java UDF and native UDF.
In the future, we should consider to provide a Doris Java UDF interface in order to support more complex types, such as HLL, Date, etc.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**